### PR TITLE
initial ts setup. changed pages/*.js to pages/*.tsx

### LINKED ISSUE #1721: Switch to TypeScript

### PR DESCRIPTION
**Pull request by billy-le**
# Description of changes
introduce typescript with initial setup. pages dir files converted to typescript.

# Issue Resolved

Fixes #1721

### CODE REVIEW

**Comment by billy-le** [comment]
Not sure what the scope of this issue is but I think we should add typescript incrementally. Here I just set up an initial setup with a few changes to the pages dir. We can open up other issues to convert Components since those are most likely to take the longer to transition to ts.

File: pages/_app.tsx
Comment (kylemh) [inline pages/_app.tsx:42]: https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492

```suggestion
function Layout({ children }: PropsWithChildren<unknown>) {
```

File: pages/confirm_email.tsx
Comment (kylemh) [inline pages/confirm_email.tsx:10]: Can you remove propTypes definitions for converted files and the ESLint rules around propTypes as part of this PR?

File: pages/_app.tsx
Comment (kylemh) [inline pages/_app.tsx:77]: You can remove these comments from files where you've converted the type.

File: pages/_app.tsx
Comment (kylemh) [inline pages/_app.tsx:6]:
```suggestion
import { useEffect } from 'react';
import type { PropsWithChildren } from 'react';
```

Comment (kylemh) [reply pages/_app.tsx:6]: When you enable TypeScript ESLint as requested in a later comment, [this rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-imports.md) (in it's default config) will automate this distinction.

**Review by kylemh — COMMENTED** [review]
Great work so far! I agree incremental is best.

Two things:
1. Are you able to setup strict type-checking of just `.ts,.tsx` files as part of CI. As we incrementally add TS, it'd be best if the already-converted files dont regress in type-safety.
2. Will you please setup the TypeScript ESLint plugin as part of this PR?

More on 2:
Other people may contribute in between the conversion work so it'll be good to have the ESLint plugin to guide good behavior.

I'd like to extend the `strict` and `stylistic` [presets](https://typescript-eslint.io/linting/configs/#recommended-configurations). I also want to add these custom rules:

```
'@typescript-eslint/consistent-type-imports': 'error',
'@typescript-eslint/explicit-module-boundary-types': 'off',
'@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'variable',
        types: ['boolean'],
        format: ['PascalCase', 'UPPER_CASE'],
        prefix: [
          'is',
          'was',
          'should',
          'has',
          'can',
          'did',
          'will',
          'IS_',
          'WAS_',
          'SHOULD_',
          'HAS_',
          'CAN_',
          'DID_',
          'WILL_',
        ],
      },
    ],
'@typescript-eslint/no-empty-interface': ['error', { 'allowSingleExtends': true }],
'@typescript-eslint/no-explicit-any': 'off', // we've already begun using `any` 🤷‍♂️
'@typescript-eslint/no-unused-vars': ['error', { vars: 'all', varsIgnorePattern: '_' }],
'@typescript-eslint/unbound-method': 'off', // gives false negatives in arrow funcs
```

File: pages/_app.tsx
Comment (billy-le) [reply pages/_app.tsx:6]: Would you like me to change the parser from `@babel/eslint-parser` to `@typescript-eslint/parser` too?

Comment (kylemh) [reply pages/_app.tsx:6]: We can do 2 sets of parsers! I'd keep `@babel/eslint-parser` for JS files and the TS ESLint parser for TS.

File: pages/join/form.tsx
Comment (kylemh) [inline pages/join/form.tsx:36]: check out docs on Next.js and TypeScript. There should be a type that we can verify via:

`UpdateProfile.getInitialProps: SomeTypeHere` and then everything else will work fine. I forget the name of it!

File: .eslintrc.js
Comment (kylemh) [inline .eslintrc.js:25]: you wouldn't care about tsconfig here for JS files, no?

File: .eslintrc.js
Comment (kylemh) [inline .eslintrc.js:50]: i think you just need
`project: true`

File: .eslintrc.js
Comment (kylemh) [inline .eslintrc.js:226]: these rules would need to go within overrides with the parser, no?

File: .eslintrc.js
Comment (billy-le) [reply .eslintrc.js:50]: At first, I had this parserOptions in the overrides with project set to true or the tsconfig but it kept complaining that it couldn't find the parserServices when linting.

File: .eslintrc.js
Comment (billy-le) [reply .eslintrc.js:226]: That's what I thought and put it there but for some reason it wasn't working for me. Maybe my config was bad and I'll try again.

**Comment by billy-le** [comment]
Most of the issues are now resolved

image

Let me know if there is anything else you would like for me to include.

**Comment by kylemh** [comment]
don't forget to remove prop types from all the files you converted. any initial prop values can be switched to default values within the params of the function.

**Comment by billy-le** [comment]
okay, i think i got them all

**Comment by billy-le** [comment]
this is just the initial set up of typescript. i believe that migrating individual components would be a separate pr.

**Comment by recondesigns** [comment]
> this is just the initial set up of typescript. i believe that migrating individual components would be a separate pr.

Right on. I didn't want to start working on any components until the TS setup was merged in so that I could branch off that. Was curious if you were still working this, or if it was just waiting for someone to approve the merge (despite the codeclimate not passing).

**Review by kylemh — APPROVED** [review]

**Comment by kylemh** [comment]
great work @billy-le - thanks!

**Pull request merged**
